src/index.js

```diff
@@ -57,35 +56,51 @@ class ExtractTextPlugin {
 
   mergeNonInitialChunks(chunk, intoChunk, checkedChunks) {
     if (!intoChunk) {
-      checkedChunks = [];
-      chunk.getChunks().forEach((c) => {
-        if (isInitialOrHasNoParents(c)) return;
-        this.mergeNonInitialChunks(c, chunk, checkedChunks);
-      }, this);
-    } else if (checkedChunks.indexOf(chunk) < 0) {
-      checkedChunks.push(chunk);
-      chunk.forEachModule((module) => {
-        intoChunk.addModule(module);
-        module.addChunk(intoChunk);
-      });
-      chunk.getChunks().forEach((c) => {
-        if (isInitialOrHasNoParents(c)) return;
-        this.mergeNonInitialChunks(c, intoChunk, checkedChunks);
-      }, this);
+      const newCheckedChunks = [];
+
+      for (const asyncChunk of chunk.getAllAsyncChunks()) {
+        if (!asyncChunk.isOnlyInitial()) {
+          this.mergeNonInitialChunks(asyncChunk, chunk, newCheckedChunks);
+        }
+      }
+    } else if (!checkedChunks.includes(chunk)) {
+      const newCheckedChunks = checkedChunks.concat(chunk);
+
+      for (const chunkModule of chunk.modulesIterable) {
+        intoChunk.addModule(chunkModule);
+        chunkModule.addChunk(intoChunk);
+      }
+
+      for (const asyncChunk of chunk.getAllAsyncChunks()) {
+        if (!asyncChunk.isOnlyInitial()) {
+          this.mergeNonInitialChunks(asyncChunk, intoChunk, newCheckedChunks);
+        }
+      }
     }
   }
 
-  renderExtractedChunk(chunk) {
+  static renderExtractedChunk(chunk) {
     const source = new ConcatSource();
-    chunk.forEachModule((module) => {
-      const moduleSource = module.source();
+
+    for (const chunkModule of chunk.modulesIterable) {
+      let moduleSource = chunkModule.source();
+
+      if (moduleSource instanceof CachedSource) {
```

Comment:
During my testing I noticed that async imports are CachedSource instances caching a ReplaceSource instance, which breaks the plugin.

The issue probably lies elsewhere.

Reply:
Please add a comment for this to the code :)